---
title: "Calculating MIRACLE Score from Bulk Gene Expression Datasets"
author: "Tolga Turan"
date: "June 14, 2019"
output: github_document
---

##INTRODUCTION:
**MIRACLE** is short for "**M**ediators of **I**mmune **R**esponse **A**gainst **C**ancer in so**L**id micro**E**nvironments".
The genesets are generated using a straigthforward approach as explained in the associated publication. Briefly, the cohorts where immune infiltrate has good prognosis association as well as the ones with opposite prognosis association are used as training sets.
The markers generated by the training samples are used to calculate single-sample GeneSet Enrichment (ssGSE) scores and mathematical ratio of ssGSE scores of good prognosis cohorts to worse prognosis cohorts gives Miracle Score.


##INSTALLATION:
First install the latest version of the dependency package "yaGST":

``` {r install_yaGST, echo=TRUE, eval=FALSE}
library(devtools)
install_github("miccec/yaGST")
```
Then install Miracle as follows:
``` {r install_MIRACLE, echo=TRUE, eval=FALSE}
library(devtools)
install_github("turantx1/Miracle", host="pig.abbvienet.com/api/v3", 
auth_token="cff1778f23dc72126f68e18c39b85bc853d35fc8")
```

##USAGE:
To calculate Miracle score, use a gene expression matrix normalized by methods including but not limited to EDASeq for RNA-seq or Quantile or RMA for Microarrays.
 
Please note that Miracle has not been tested on Nanostring platforms or RNAseq RPKM, FPKM or TPM normalizations.

The accepted feature_identifiers are "ens" for Ensemble Gene Ids; "u133p2" for Affymetrix U133p2 or U133A probe_ids; "entrez" for Entrez Gene Ids; "gene" for Gene Symbols and "ilmn" for IlluminaHT12v3 or IlluminaHT12v4 microarray platforms. 

###Example: Calculating Miracle Scores and Measuring Survival Association in TCGA-BLCA:
_Please note that this vignette makes use of "ggplot2" and "survival" packages which are not necessary for installation but necessary to replicate this vignette._

Load normalized gene expression and Survival Data for TCGA-BLCA
``` {r load_data, echo=TRUE, eval=TRUE}
library(survival)
library(ggplot2)
library(Miracle)
data("SURV_BLCA","TCGA_BLCA_EDASeq_norm",package="Miracle") 
head(SURV_BLCA)
TCGA_BLCA_EDASeq_norm[1:5, 1:5]
```

Calculate ICR and Miracle Scores and merge with survival data:

``` {r calculate_miracle, results="hide"}
TCGA_ICR_and_miracle_scores<-Calculate_Miracle(TCGA_BLCA_EDASeq_norm, "ens")
```

``` {r merge_data, echo=TRUE, eval=TRUE}
head(TCGA_ICR_and_miracle_scores)
all(rownames(TCGA_ICR_and_miracle_scores)==SURV_BLCA[,1])
TCGA_BLCA_Miracle_SURV<-cbind(SURV_BLCA, TCGA_ICR_and_miracle_scores)
head(TCGA_BLCA_Miracle_SURV)
```
 Calculate Survival association of ICR in TCGA_BLCA Cohort:

``` {r survival_association_ICR, echo=TRUE, eval=TRUE}
coxph(Surv(Time, Status)~ICR, data=TCGA_BLCA_Miracle_SURV)
```
Calculate Survival association of Miracle in TCGA_BLCA Cohort:

``` {r survival_association_Miracle, echo=TRUE, eval=TRUE}
coxph(Surv(Time, Status)~Miracle, data=TCGA_BLCA_Miracle_SURV)
```

_Similar to this example, Miracle scores calculated by 'Calculate\_Miracle' function can be merged with immunotherapy datasets with response outcomes._

 
